Initial implementation of Amplitude service

darinkrauss opened the pull request:
- Implement Service and ServiceUI protocols
- Implement Analytics protocol
- Connect with Amplitude third-party service

https://tidepool.atlassian.net/browse/LOOP-22

darinkrauss committed:
Initial implementation of Amplitude service

- Implement Service and ServiceUI protocols
- Implement Analytics protocol
- Connect with Amplitude third-party service

nhamming commented on AmplitudeServiceKit/en.lproj/Localizable.strings:
```diff
@@ -0,0 +1,2 @@
+/* The title of the Amplitude service */
+"Amplitude" = "Amplitude";
```
Maybe I'm mistaken, but won't the default language localization be provided when a `.strings` file is missing. If yes, I don't see the point of making `.strings` files that do not provide any localization.

darinkrauss replied:
Yes, good call. Will remove in the three service repos.

nhamming commented on AmplitudeServiceKitTests/AmplitudeServiceKitTests.swift:
```diff
@@ -0,0 +1,34 @@
+import XCTest
+@testable import AmplitudeServiceKit
+
+class AmplitudeServiceKitTests: XCTestCase {
```
Should this be committed?

darinkrauss replied:
Probably not. Will remove the test file itself, but leave the project.

nhamming commented on AmplitudeServiceKitUI/AmplitudeServiceTableViewController.swift:
```diff
@@ -0,0 +1,147 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        tableView.register(AuthenticationTableViewCell.nib(), forCellReuseIdentifier: AuthenticationTableViewCell.className)
```
Just want to flag the `.nib()` since I recently saw a change to `.self`.

darinkrauss replied:
The use of `.self` or `.nib()` depends upon whether the view is implemented strictly in code (where `.self` would be used) or whether it uses a `.xib` file (where `.nib()` would be used). 

In the case of the recent change from `.nib()` to `.self` for `SwitchTableViewCell`, the `.xib` file was recently deleted and the view implemented only in code. See 9b23b8330f3507aa39ccd3f716b556b033f3c8ba for details.

In the above case, the `AuthenticationTableViewCell` uses a `.xib` file so the use of `.nib()` is correct.

nhamming reviewed (dismissed):
Otherwise LGTM

darinkrauss committed:
Updates from PR review

- Remove unnecessary localizable strings file for English
- Remove unused test case files

darinkrauss committed:
Updates reflecting service protocol renaming

darinkrauss committed:
Rename Service hasValidConfiguration to hasConfiguration for PR review

darinkrauss committed:
Update service completion and notification delegate from PR review

darinkrauss committed:
Remove extraneous whitespace

nhamming reviewed (dismissed):
LGTM

darinkrauss committed:
Refactor service setup delegate from PR comments

darinkrauss committed:
Remove unnecessary elements of Service protocol

nhamming reviewed (dismissed):
+1

ps2 commented on AmplitudeServiceKitUI/AmplitudeServiceTableViewController.swift:
```diff
@@ -0,0 +1,232 @@
+    convenience init(serviceDeletionHandler handler: @escaping () -> Void) {
+        self.init(
+            title: nil,
+            message: NSLocalizedString("Are you sure you want to delete this service?", comment: "Confirmation message for deleting a service"),
```
NSLocalizedString -> LocalizedString, and a couple more below.

ps2 commented on Common/LocalizedString.swift:
```diff
@@ -0,0 +1,23 @@
+
+func LocalizedString(_ key: String, tableName: String? = nil, value: String? = nil, comment: String) -> String {
+    if let value = value {
+        return NSLocalizedString(key, tableName: tableName, bundle: FrameworkBundle.main, value: value, comment: comment)
```
NSLocalizedString -> LocalizedString, and one below.

ps2 reviewed (changes requested):
Looks good. Just some conversions of NSLocalizedString -> LocalizedString

darinkrauss committed:
Replace NSLocalizedString with LocalizeString per PR review

ps2 reviewed (approved):
🎉 LGTM!
